scipy/linalg/decomp.py

```diff
@@ -371,126 +415,170 @@ def eigh(a, b=None, lower=True, eigvals_only=False, overwrite_a=False,
+
+    # Check indices if given
+    if subset_by_index:
+        lo, hi = [int(x) for x in subset_by_index]
+        if not (0 <= lo <= hi < n):
+            raise ValueError('Requested eigenvalue indices are not valid. '
+                             'Valid range is [0, {}] and start <= end, but '
+                             'start={}, end={} is given'.format(n-1, lo, hi))
+        # fortran is 1-indexed
+        drv_args.update({'range': 'I', 'il': lo + 1, 'iu': hi + 1})
+
+    if subset_by_value:
+        lo, hi = subset_by_value
+        if not (-inf <= lo < hi <= inf):
+            raise ValueError('Requested eigenvalue bounds are not valid. '
+                             'Valid range is (-inf, inf) and low < high, but '
+                             'low={}, high={} is given'.format(lo, hi))
+
+        drv_args.update({'range': 'V', 'vl': lo, 'vu': hi})
 
     # fix prefix for lapack routines
-    if cplx:
-        pfx = 'he'
+    pfx = 'he' if cplx else 'sy'
+
+    # decide on the driver if not given
+    # first early exit on incompatible choice
+    if driver:
+        if b is None and (driver in ["gv", "gvd", "gvx"]):
+            raise ValueError('{} requires input b array to be supplied '
+                             'for generalized eigenvalue problems.'
+                             ''.format(driver))
+        if (b is not None) and (driver in ['ev', 'evd', 'evr', 'evx']):
+            raise ValueError('"{}" does not accept input b array '
+                             'for standard eigenvalue problems.'
+                             ''.format(driver))
+        if subset and (driver in ["ev", "evd", "gv", "gvd"]):
+            raise ValueError('"{}" cannot compute subsets of eigenvalues'
+                             ''.format(driver))
+
+    # Default driver is evr and gvd
     else:
-        pfx = 'sy'
-
-    #  Standard Eigenvalue Problem
-    #  Use '*evr' routines
-    # FIXME: implement calculation of optimal lwork
-    #        for all lapack routines
-    if b1 is None:
-        driver = pfx+'evr'
-        (evr,) = get_lapack_funcs((driver,), (a1,))
-        if eigvals is None:
-            w, v, info = evr(a1, uplo=uplo, jobz=_job, range="A", il=1,
-                             iu=a1.shape[0], overwrite_a=overwrite_a)
+        driver = "evr" if b is None else ("gvx" if subset else "gvd")
+
+    lwork_spec = {
+                  'syevd': ['lwork', 'liwork'],
+                  'syevr': ['lwork', 'liwork'],
+                  'heevd': ['lwork', 'liwork', 'lrwork'],
+                  'heevr': ['lwork', 'liwork', 'lrwork'],
+                  }
+
+    if b is None:  # Standard problem
+        drv, drvlw = get_lapack_funcs((pfx + driver, pfx+driver+'_lwork'),
+                                      [a1])
+        lw = _compute_lwork(drvlw, n, lower=lower)
+        # Multiple lwork vars
+        if isinstance(drvlw, tuple):
+            lwork_args = {x: y for x, y in zip(lwork_spec[pfx+driver], lw)}
```

Comment:
I think:
```suggestion
            lwork_args = dict(zip(lwork_spec[pfx+driver], lw)
```

Reply:
Nice! I should remember this more often.